src/test_application/activescenes/scenarios_universetest.cpp

```diff
@@ -0,0 +1,701 @@
+
+    auto const physIn = ArrayView<ACtxPhysInputs>(&rScnPhys.m_physIn, 1);
+    SysNewton::update_world(
+            rScnPhys.m_physics, rScnNwt, delta, physIn,
+            rScene.m_basic.m_hierarchy,
+            rScene.m_basic.m_transform, rScene.m_basic.m_transformControlled,
+            rScene.m_basic.m_transformMutable);
+
+    // Start recording new elements to delete
+    rScene.m_delete.clear();
+
+    // Check position of all entities with the out-of-bounds component
+    // Delete the ones that go out of bounds
+    for (ActiveEnt const ent : rScnTest.m_removeOutOfBounds)
+    {
+        ACompTransform const &entTf = rScene.m_basic.m_transform.get(ent);
+        if (entTf.m_transform.translation().y() < -10)
+        {
+            rScene.m_delete.push_back(ent);
+        }
+    }
+
+    // Delete entities in m_delete, their descendants, and components
+    update_test_scene_delete(rScene);
+
+    // Shape Thrower system, consumes rScene.m_toThrow and creates shapes
+    for (UniverseTestData::ThrowShape const &rThrow : std::exchange(rScnTest.m_toThrow, {}))
+    {
+        ActiveEnt shapeEnt = add_rigid_body_quick(
+                rScene, rThrow.m_position, rThrow.m_velocity, rThrow.m_mass,
+                rThrow.m_shape, rThrow.m_size);
+
+        // Make gravity affect entity
+        rScnTest.m_hasGravity.emplace(shapeEnt);
+
+        // Remove when it goes out of bounds
+        rScnTest.m_removeOutOfBounds.emplace(shapeEnt);
+    }
+
+    // Sort hierarchy, required by renderer
+    SysHierarchy::sort(rScene.m_basic.m_hierarchy);
+}
+
+
+static void update_universe(CommonTestScene& rScene, float delta)
+{
+    using namespace osp::universe;
+
+    auto &rScnTest = rScene.get<UniverseTestData>();
+    auto &rScnPhys = rScene.get<PhysicsData>();
+    Universe &rUni = rScnTest.m_universe;
+    CoSpaceCommon &rMainSpace = rUni.m_coordCommon[std::size_t(rScnTest.m_mainSpace)];
+
+    float const scale = osp::math::mul_2pow<float, int>(1.0f, -rMainSpace.m_precision);
+    float const scaleDelta = delta / scale;
+
+    auto const [x, y, z]        = sat_views(rMainSpace.m_satPositions, rMainSpace.m_data, rMainSpace.m_satCount);
+    auto const [vx, vy, vz]     = sat_views(rMainSpace.m_satVelocities, rMainSpace.m_data, rMainSpace.m_satCount);
+    auto const [qx, qy, qz, qw] = sat_views(rMainSpace.m_satRotations, rMainSpace.m_data, rMainSpace.m_satCount);
+
+    // Phase 1: Move satellites
+
+    for (std::size_t i = 0; i < rMainSpace.m_satCount; ++i)
+    {
+        x[i] += vx[i] * scaleDelta;
+        y[i] += vy[i] * scaleDelta;
+        z[i] += vz[i] * scaleDelta;
+
+        // Apply arbitrary inverse-square gravity towards origin
+        Vector3d const pos       = Vector3d( Vector3g( x[i], y[i], z[i] ) ) * scale;
+        float const r           = pos.length();
+        float const c_gm        = 10000000000.0f;
+        Vector3d const accel     = -pos * delta * c_gm / (r * r * r);
```

Comment:
Arithmetic overflow: Using operator '' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '' to avoid overflow (io.2).

Warning
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
Show more details